### PR TITLE
Add SCLF-0001: JSON Key Strategies

### PR DESCRIPTION
**Pull request by parkera**
Add a proposal for handling snake and camel case keys in JSONEncoder.

### CODE REVIEW

File: Docs/Proposals/0001-jsonencoder-key-strategy.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+Previously, you would customize your `Decodable` type with custom keys, like this:
+
+```swift
+struct Thing : Decodable {
+
+    let helloWorld : Int
+    let goodbyeCruelWorld: Int
+    let key: Int
+
+    private enum CodingKeys : CodingKey {
```
Comment (ianpartridge) [inline Docs/Proposals/0001-jsonencoder-key-strategy.md]: This should have a raw type of `String`, I think?

Comment (parkera) [reply Docs/Proposals/0001-jsonencoder-key-strategy.md]: Yup

File: Docs/Proposals/0001-jsonencoder-key-strategy.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+1. Capitalizes the word starting after each `_`
+2. Removes all `_`
+3. Preserves starting and ending `_` (as these are often used to indicate private variables or other metadata).
+
+For example, `one_two_three` becomes `oneTwoThree`. `_one_two_three_` becomes `_oneTwoThree_`.
+
+Converting from camel case to snake case:
+
+1. Splits words at the boundary of lower-case to upper-case
```
Comment (bubski) [inline Docs/Proposals/0001-jsonencoder-key-strategy.md]: If I undestand correctly, this implies `imageURL` becomes `image_URL`, right? Or rather `image_url` I guess.
Can we add examples with acronyms to make it clear?

Comment (bubski) [reply Docs/Proposals/0001-jsonencoder-key-strategy.md]: ah, right
>Lowercases the entire string
so `image_url` would be the one, yes?

File: Docs/Proposals/0001-jsonencoder-key-strategy.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+2. Inserts `_` between words
+3. Lowercases the entire string
+4. Preserves starting and ending `_`.
```
Comment (bubski) [inline Docs/Proposals/0001-jsonencoder-key-strategy.md]: How would we approach numbers? 
E.g. what would `abc123def` be resolved to?

File: Docs/Proposals/0001-jsonencoder-key-strategy.md (hunk 2 above)
Comment (parkera) [reply Docs/Proposals/0001-jsonencoder-key-strategy.md]: Yah, check out the test cases at https://github.com/apple/swift/pull/12779/ for some additional examples.

File: Docs/Proposals/0001-jsonencoder-key-strategy.md (hunk 3 above)
Comment (parkera) [reply Docs/Proposals/0001-jsonencoder-key-strategy.md]: `abc123_def`

Comment (parkera) [reply Docs/Proposals/0001-jsonencoder-key-strategy.md]: `abc123Def` would be `abc123_def`

**Comment by ianpartridge** [comment]
@parkera do you want to merge this or leave it as a PR?

**Comment by parkera** [comment]
I think I'll go ahead and merge this now.

**Pull request merged**